# Language Model
**Instructions**:
1. Wherever you are asked to code, insert a text block below your code block and explain what you have coded as per your own understanding.
2. If the code is provided by us, execute it, and add below a text block and provide your explanation in your own words.
3. Submit both the .ipynb and pdf files to canvas.
4. **The similarity score should be less than 15%.**

# Task 1 (10 points)
Explain the importance of Ngram language modeling? Provide an example of a practical application of Ngram language modeling in industry.

**Answer here***
**Language model:**
Language model is used for calculating the probablity of the sequences of the words based on conditional and joint probability.
There are two types of the language model:

1.statistical language model:
Statistical language model is use the traditional techniques to compute the probability of the words based on the markov decision assumption adn Ngram techniques.
2.Neural language model:
Neural language model is use different kind of natural language models for language processing.

There are different kind of gram language models are avialable that are listed below:Focusing on some of the main gram language models are:
1.Unigram language model
2.bigram language model
3.Ngram language model
Now, takling abput all the models in details we first focus on the first model,

**1.unigram language model:**
The unigram language model is basically calculating the probability of the words without condition its calculated independently.In short the unigram language model is only give the probability of the 1 word individually.
for instance,the sentence is "I like to eat pizza".then, talking about the results given by the unigram model is "I","like","to","eat","pizza".its give each words probability indivually and not checking the conditional probability of other words of the sentences.
**2.bigram language model**
The bigram language model is basically calculating the probability of the two words based on condition. its consider two sequence of the words and give the probability of the occurance.For instances,the sentence is "we are in class right now". the results provide by bigram is like"We are"","in class","right now",so all in all we are getting the probability of the two sequences of the words using the conditional probability.its gives atleast better result apart from the unigram.
**3.Ngram language model**
Talking about the Ngram language model this model is used for calculating conditional probability of the last word based on the previously occures words. For training this model they use corpus of text.Ngram language model is the sequence of number of words.
p(w)=p(wi|wi-1)
where, wi stands for last word of the sentece.
wi-1 stands for last to second word of the sentence
basically, Ngram model is usinf one or two words of the sentence and getting the probability of the language model.
**There are some application which is used Ngram language model:**
Ngram model is now majorly used for probability , communication theory,calculation statical natural language processing, biological sequence analysis and data compression.
Application name of the Ngram language model used.
Complition of sentences.

For instance, text analytics,companies like Google, Amazon and Microsoft use ngram language model to develop their search engines,which is based on the ability for predict the search queries and generate matching results.Ngram is use for machine translation, where they are using to predict the translation of words and phrases based on their content.

Focusing on the second example, using ngram language model is in predictive text applications like mobile phone keyboards and messaging apps. that model predict the words and phrase the typing based on the context of the previous words.

Insummary, Ngram language model is an important technique in natrual language model is used to predict the next word in a sequence of words.Its has lots of practical application including search engines, machine translation and predicting text.

# Tutorial 1: N-grams Language Model

This tutorial serves as an example for creating your own Python language model. The code given Below is a  basic language model using trigrams of the Reuters corpus. A language model is a form of artificial neural network that has been trained to estimate, given some input text, the likelihood of a sequence of words or characters. The fundamental language model implies that it predicts the character that will appear after each character that has come before it.

An n-gram model is a common example of a probabilistic language model. It entails estimating the likelihood of the following word in a phrase given the preceding n-1 words. For instance, a 2-gram (often referred to as a bigram) model would forecast the likelihood of the following word given the preceding word.

Additionally, the tutorial below offers thorough code examples for each stage, making it simple for beginners to follow along and create their own Python language model. You should have a fundamental understanding of language modeling by the end of the lesson and be able to use this knowledge for language modeling Problems.

https://medium.com/analytics-vidhya/a-comprehensive-guide-to-build-your-own-language-model-in-python-5141b3917d6d

In [ ]:
# code courtesy of https://nlpforhackers.io/language-models/

import nltk
nltk.download('reuters')
nltk.download('punkt')

from nltk.corpus import reuters
from nltk import bigrams, trigrams
from collections import Counter, defaultdict

# Create a placeholder for model
reutersmodel = defaultdict(lambda: defaultdict(lambda: 0))

# Count frequency of co-occurance
for sentence in reuters.sents():
    for w1, w2, w3 in trigrams(sentence, pad_right=True, pad_left=True):
        reutersmodel[(w1, w2)][w3] += 1

# Let's transform the counts to probabilities
for w1_w2 in reutersmodel:
    total_count = float(sum(reutersmodel[w1_w2].values()))
    for w3 in reutersmodel[w1_w2]:
        reutersmodel[w1_w2][w3] /= total_count

[nltk_data] Downloading package reuters to /root/nltk_data...
[nltk_data]   Package reuters is already up-to-date!
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


Here, we are using dataset for generating trigram basic language model.The main aim is to do text summarization  for which is used for calculating probability of the last word based on previous words and pridict three sequences of words.

import nltk - nltk stands for natural language toolkit. which is used for processing data created by human language data.

nltk.download('reuters')- this is library for dataset.this dataset is the collection of the news articles related data.

nltk.download('punkt')-this library is used for divide the paragraphs or long text into the list of sentences using the unsupervised algorithm for building the  model with abbreviation words, collocations and starting sentences words.

from nltk.corpus import reuters - this is news corpus dataset which is used for creating basic language model using trigram language model method.

from nltk import bigrams, trigrams- this are the libraries for making language model for speech recognitions. bigram is used for calculating the probability of words based on thired word with previous two words its return the two words sequences. However, trigram is calculating the probability of the words based on forth word with conditon of prevoius words it will give the result of three continues words sequences.

from collections import Counter, defaultdict - This is used for collection.counter is a dictionary subclass for counting the hashable object.this is collection of elements keys and their counts are stores as dictionary values.
collection.defaultdict is the subclass its used for specify default value for nonexistance key, if the data is not there then if its not give error insted of return the default values.

model = defaultdict(lambda: defaultdict(lambda: 0))- here we are creating the placeholder for model.

For loop  - we are using for loop for counting the frequency of the words occurance.

calculating the probability - here we are making the for loop for calculating the words probabilities.

In [ ]:
# code courtesy of https://nlpforhackers.io/language-models/

import random

# starting words
text = ["today", "the"]
sentence_finished = False

while not sentence_finished:
  # select a random probability threshold
  r = random.random()
  accumulator = .0

  for word in reutersmodel[tuple(text[-2:])].keys():
      accumulator += reutersmodel[tuple(text[-2:])][word]
      # select words that are above the probability threshold
      if accumulator >= r:
          text.append(word)
          break

  if text[-2:] == [None, None]:
      sentence_finished = True

print (' '.join([t for t in text if t]))

today the company .


Here, we are finding the probability of the words based on the language model and finiding the sentence based on the two first starting words.

import random - this is library which is used for generating the random integer and its random module.

text = ["today", "the"] - here, we are pre-difining the specific starting word and search the sentence which is starting from this two words.

After that we are using the constant value as a sentenced finished with false and then we are applying the threshold for random probability which is .0. additionally, we are combining all the items in to one variable using tuple now, we are selecting words based on the above threshold of the probability now we are appending the words into the text using the if condition for threshold.

now we are use if condition based on text and set the constant value sentence_finished as a true.

At the last , we are joining the sentence using the t for t in text if sentence start with t then return the sentence using two staring specified words.

In conclude, we are usign the basic language model for finding the correct sentence based on the providing two words as a condition for fetch the sentence which is start with this two words.

# Task 2 ( 30 points)
Choose a data set of your own choice and try building your own basic language model. You can refer to the tutorial in the provided above. Provide your understanding and explain in detail in a new text box below.

In [ ]:
# code courtesy of https://nlpforhackers.io/language-models/

import nltk
nltk.download('gutenberg')
nltk.download('punkt')

from nltk.corpus import gutenberg
from nltk import bigrams, trigrams
from collections import Counter, defaultdict

# Create a placeholder for model
gutenbergmodel = defaultdict(lambda: defaultdict(lambda: 0))

# Count frequency of co-occurance
for sentence in gutenberg.sents():
    for w1, w2, w3 in trigrams(sentence, pad_right=True, pad_left=True):
        gutenbergmodel[(w1, w2)][w3] += 1

# Let's transform the counts to probabilities
for w1_w2 in gutenbergmodel:
    total_count = float(sum(gutenbergmodel[w1_w2].values()))
    for w3 in gutenbergmodel[w1_w2]:
        gutenbergmodel[w1_w2][w3] /= total_count

[nltk_data] Downloading package gutenberg to /root/nltk_data...
[nltk_data]   Package gutenberg is already up-to-date!
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


Here, we are using dataset for generating trigram basic language model.The main aim is to do text summarization for which is used for calculating probability of the last word based on previous words and pridict three sequences of words.

import nltk - nltk stands for natural language toolkit. which is used for processing data created by human language data.

nltk.download('gutenberg')- this is library for dataset.this datset is used for statistical analysis of the language.this dataset is the collection of the electronics books.

nltk.download('punkt')-this library is used for divide the paragraphs or long text into the list of sentences using the unsupervised algorithm for building the model with abbreviation words, collocations and starting sentences words.

from nltk.corpus import gutenberg - this is book corpus dataset which is used for creating basic language model using trigram language model method.

from nltk import bigrams, trigrams- this are the libraries for making language model for speech recognitions. bigram is used for calculating the probability of words based on thired word with previous two words its return the two words sequences. However, trigram is calculating the probability of the words based on forth word with conditon of prevoius words it will give the result of three continues words sequences.

from collections import Counter, defaultdict - This is used for collection.counter is a dictionary subclass for counting the hashable object.this is collection of elements keys and their counts are stores as dictionary values. collection.defaultdict is the subclass its used for specify default value for nonexistance key, if the data is not there then if its not give error insted of return the default values.

model = defaultdict(lambda: defaultdict(lambda: 0))- here we are creating the placeholder for model.

For loop - we are using for loop for counting the frequency of the words occurance.

calculating the probability - here we are making the for loop for calculating the words probabilities.

In [ ]:
# code courtesy of https://nlpforhackers.io/language-models/

import random

# starting words
text = ["today", "the"]
sentence_finished = False

while not sentence_finished:
  # select a random probability threshold
  r = random.random()
  accumulator = .0

  for word in gutenbergmodel[tuple(text[-2:])].keys():
      accumulator += gutenbergmodel[tuple(text[-2:])][word]
      # select words that are above the probability threshold
      if accumulator >= r:
          text.append(word)
          break

  if text[-2:] == [None, None]:
      sentence_finished = True

print (' '.join([t for t in text if t]))

today the conquering march of science , is but perceptibly open ' d with frantic shouts , crash of glass .


**Answer here:**
Here, we are finding the probability of the words based on the language model and finiding the sentence based on the two first starting words.

import random - this is library which is used for generating the random integer and its random module.

text = ["today", "the"] - here, we are pre-difining the specific starting word and search the sentence which is starting from this two words.

After that we are using the constant value as a sentenced finished with false and then we are applying the threshold for random probability which is .0. additionally, we are combining all the items in to one variable using tuple now, we are selecting words based on the above threshold of the probability now we are appending the words into the text using the if condition for threshold.

now we are use if condition based on text and set the constant value sentence_finished as a true.

At the last , we are joining the sentence using the t for t in text if sentence start with t then return the sentence using two staring specified words.

In conclude, we are usign the basic language model for finding the correct sentence based on the providing two words as a condition for fetch the sentence which is start with this two words.

# Task 3 (10 points)
Explain the steps involved in building a neural language model for natural language processing (NLP)? Refer to the tutorial provided above.

**Answer here:**
Talking about the article its majorly talking about the text summarization and creating the language model for that which is predict the next words based on the current and previous words or we can say auto complete the next predict word of the sentences.
Focusing on the first part, they are talking about the language model as per their opinion the language model is trained to predict the probability of the sequences of the words.

There are two different types of the language model:
1. Statical language model :  This model is use for the traditional statistical techniques such as Ngram,markov model learn to probability of the words.
2. Neural network model : this technique use different kind of neural network models.

Talking about the neural network model we are basically making two different types of model character level and word level.

**Step-1**
Here we are importing libraries for creating the language model using neural network language model.

In [ ]:
import numpy as np
import pandas as pd
from keras.utils import to_categorical
from keras import preprocessing
from keras.preprocessing import sequence
from tensorflow.keras.preprocessing.sequence import pad_sequences

#!pip install keras_preprocessing.sequence.pad_sequences
#import pad_sequences
#from keras.preprocessing.sequence import pad_sequences
from keras.models import Sequential
from keras.layers import LSTM, Dense, GRU, Embedding
from keras.callbacks import EarlyStopping, ModelCheckpoint

Here,we are importing the libraries such as,
numpy - for processing or analytics the text.
pandas - for performing statistical analysis of text.
keras - keras is well known library for developing the deep learnign models.here, we are going to create sequential model from keras.models.in short here for creating neural language model we are going to use the keras model  and also we are going to use the multiple layers for the model which are lstm,dense.GRU and embedding. apart from that we are using call back library for easily stopping and model checkpoints. In short, we are importing the liabraries for generating the neural language model.

**Step-2** Read the dataset.Here we are just reading the provided data.

In [ ]:
data_text = """The unanimous Declaration of the thirteen united States of America, When in the Course of human events, it becomes necessary for one people to dissolve the political bands which have connected them with another, and to assume among the powers of the earth, the separate and equal station to which the Laws of Nature and of Nature's God entitle them, a decent respect to the opinions of mankind requires that they should declare the causes which impel them to the separation.

We hold these truths to be self-evident, that all men are created equal, that they are endowed by their Creator with certain unalienable Rights, that among these are Life, Liberty and the pursuit of Happiness.--That to secure these rights, Governments are instituted among Men, deriving their just powers from the consent of the governed, --That whenever any Form of Government becomes destructive of these ends, it is the Right of the People to alter or to abolish it, and to institute new Government, laying its foundation on such principles and organizing its powers in such form, as to them shall seem most likely to effect their Safety and Happiness. Prudence, indeed, will dictate that Governments long established should not be changed for light and transient causes; and accordingly all experience hath shewn, that mankind are more disposed to suffer, while evils are sufferable, than to right themselves by abolishing the forms to which they are accustomed. But when a long train of abuses and usurpations, pursuing invariably the same Object evinces a design to reduce them under absolute Despotism, it is their right, it is their duty, to throw off such Government, and to provide new Guards for their future security.--Such has been the patient sufferance of these Colonies; and such is now the necessity which constrains them to alter their former Systems of Government. The history of the present King of Great Britain is a history of repeated injuries and usurpations, all having in direct object the establishment of an absolute Tyranny over these States. To prove this, let Facts be submitted to a candid world.

He has refused his Assent to Laws, the most wholesome and necessary for the public good.

He has forbidden his Governors to pass Laws of immediate and pressing importance, unless suspended in their operation till his Assent should be obtained; and when so suspended, he has utterly neglected to attend to them.

He has refused to pass other Laws for the accommodation of large districts of people, unless those people would relinquish the right of Representation in the Legislature, a right inestimable to them and formidable to tyrants only.

He has called together legislative bodies at places unusual, uncomfortable, and distant from the depository of their public Records, for the sole purpose of fatiguing them into compliance with his measures.

He has dissolved Representative Houses repeatedly, for opposing with manly firmness his invasions on the rights of the people.

He has refused for a long time, after such dissolutions, to cause others to be elected; whereby the Legislative powers, incapable of Annihilation, have returned to the People at large for their exercise; the State remaining in the mean time exposed to all the dangers of invasion from without, and convulsions within.

He has endeavoured to prevent the population of these States; for that purpose obstructing the Laws for Naturalization of Foreigners; refusing to pass others to encourage their migrations hither, and raising the conditions of new Appropriations of Lands.

He has obstructed the Administration of Justice, by refusing his Assent to Laws for establishing Judiciary powers.

He has made Judges dependent on his Will alone, for the tenure of their offices, and the amount and payment of their salaries.

He has erected a multitude of New Offices, and sent hither swarms of Officers to harrass our people, and eat out their substance.

He has kept among us, in times of peace, Standing Armies without the Consent of our legislatures.

He has affected to render the Military independent of and superior to the Civil power.

He has combined with others to subject us to a jurisdiction foreign to our constitution, and unacknowledged by our laws; giving his Assent to their Acts of pretended Legislation:

For Quartering large bodies of armed troops among us:

For protecting them, by a mock Trial, from punishment for any Murders which they should commit on the Inhabitants of these States:

For cutting off our Trade with all parts of the world:

For imposing Taxes on us without our Consent:

For depriving us in many cases, of the benefits of Trial by Jury:

For transporting us beyond Seas to be tried for pretended offences

For abolishing the free System of English Laws in a neighbouring Province, establishing therein an Arbitrary government, and enlarging its Boundaries so as to render it at once an example and fit instrument for introducing the same absolute rule into these Colonies:

For taking away our Charters, abolishing our most valuable Laws, and altering fundamentally the Forms of our Governments:

For suspending our own Legislatures, and declaring themselves invested with power to legislate for us in all cases whatsoever.

He has abdicated Government here, by declaring us out of his Protection and waging War against us.

He has plundered our seas, ravaged our Coasts, burnt our towns, and destroyed the lives of our people.

He is at this time transporting large Armies of foreign Mercenaries to compleat the works of death, desolation and tyranny, already begun with circumstances of Cruelty & perfidy scarcely paralleled in the most barbarous ages, and totally unworthy the Head of a civilized nation.

He has constrained our fellow Citizens taken Captive on the high Seas to bear Arms against their Country, to become the executioners of their friends and Brethren, or to fall themselves by their Hands.

He has excited domestic insurrections amongst us, and has endeavoured to bring on the inhabitants of our frontiers, the merciless Indian Savages, whose known rule of warfare, is an undistinguished destruction of all ages, sexes and conditions.

In every stage of these Oppressions We have Petitioned for Redress in the most humble terms: Our repeated Petitions have been answered only by repeated injury. A Prince whose character is thus marked by every act which may define a Tyrant, is unfit to be the ruler of a free people.

Nor have We been wanting in attentions to our Brittish brethren. We have warned them from time to time of attempts by their legislature to extend an unwarrantable jurisdiction over us. We have reminded them of the circumstances of our emigration and settlement here. We have appealed to their native justice and magnanimity, and we have conjured them by the ties of our common kindred to disavow these usurpations, which, would inevitably interrupt our connections and correspondence. They too have been deaf to the voice of justice and of consanguinity. We must, therefore, acquiesce in the necessity, which denounces our Separation, and hold them, as we hold the rest of mankind, Enemies in War, in Peace Friends.

We, therefore, the Representatives of the united States of America, in General Congress, Assembled, appealing to the Supreme Judge of the world for the rectitude of our intentions, do, in the Name, and by Authority of the good People of these Colonies, solemnly publish and declare, That these United Colonies are, and of Right ought to be Free and Independent States; that they are Absolved from all Allegiance to the British Crown, and that all political connection between them and the State of Great Britain, is and ought to be totally dissolved; and that as Free and Independent States, they have full Power to levy War, conclude Peace, contract Alliances, establish Commerce, and to do all other Acts and Things which Independent States may of right do. And for the support of this Declaration, with a firm reliance on the protection of divine Providence, we mutually pledge to each other our Lives, our Fortunes and our sacred Honor."""

**Step-3** pre-process the data

In [ ]:
import re

def text_cleaner(text):
    # lower case text
    newString = text.lower()
    newString = re.sub(r"'s\b","",newString)
    # remove punctuations
    newString = re.sub("[^a-zA-Z]", " ", newString)
    long_words=[]
    # remove short word
    for i in newString.split():
        if len(i)>=3:
            long_words.append(i)
    return (" ".join(long_words)).strip()

# preprocess the text
data_new = text_cleaner(data_text)

Here, we are importing the re which stands for regular expression. now creating the function name as a text cleaner where we passed the parameter as text.now we are converting text into lower case and now removing punctuations fro  the text using the condition after that we are removing short length words from the string.after that at the last we are joining the words and making a string. in the last we are cleaning the data using text clarner.

The first step is completed the preprocess data for generating the neural language model.

**Step-4**
Here, now we are creating the sequence.

In [ ]:
def create_seq(text):
    length = 30
    sequences = list()
    for i in range(length, len(text)):
        # select sequence of tokens
        seq = text[i-length:i+1]
        # store
        sequences.append(seq)
    print('Total Sequences: %d' % len(sequences))
    return sequences

# create sequences
sequences = create_seq(data_new)

Total Sequences: 7052


Here, we are creating the sequence of the word a from the text and providing length as 30 and making list of the sequences.
Additionally, we are creating for loop and traversing up to the length of the text and selecting the sequences of the word tokens and we are appending the tokens and finally we are printing the total sequence length of the text which is 7052. at the last we are creating object for sequences.

At the last we are getting the sequence of the words list and length of the sequences for model.

**Step-5** Encoding the sequences

In [ ]:
# create a character mapping index
chars = sorted(list(set(data_new)))
mapping = dict((c, i) for i, c in enumerate(chars))

def encode_seq(seq):
    sequences = list()
    for line in seq:
        # integer encode line
        encoded_seq = [mapping[char] for char in line]
        # store
        sequences.append(encoded_seq)
    return sequences

# encode the sequences
sequences = encode_seq(sequences)

Here, we are encoding the character of the sequence words.
first we sorte the new sequence of words and now mapping the words using the character and list of sequences. at the last we are encoding the sequences.

At the last encoding the each words of the text into the integer number of encoding.

**Step-6** Now we are creating the training and validation set

In [ ]:
from sklearn.model_selection import train_test_split

# vocabulary size
vocab = len(mapping)
sequences = np.array(sequences)
# create X and y
X, y = sequences[:,:-1], sequences[:,-1]
# one hot encode y
y = to_categorical(y, num_classes=vocab)
# create train and validation sets
X_tr, X_val, y_tr, y_val = train_test_split(X, y, test_size=0.1, random_state=42)

print('Train shape:', X_tr.shape, 'Val shape:', X_val.shape)

Train shape: (6346, 30) Val shape: (706, 30)


Here, first we are importing some libraries from the train test split library. we are first checking the size of the vocabulary length using len() after that we are creating object for sequences of array. after that we are creating x and y and devide the sequences into two parts. after that we start doing the train and validation set where we we are spliting the data using the train test split.
here we are deviding data into 90% to 10% and at the last we are printing the train shape and validation shape where the train shape is (6346,30) and validation shape is (706,30).

**Step-7** we are creating a model

In [ ]:
# define model
model = Sequential()
model.add(Embedding(vocab, 50, input_length=30, trainable=True))
model.add(GRU(150, recurrent_dropout=0.1, dropout=0.1))
model.add(Dense(vocab, activation='softmax'))
print(model.summary())

# compile the model
model.compile(loss='categorical_crossentropy', metrics=['acc'], optimizer='adam')
# fit the model
model.fit(X_tr, y_tr, epochs=100, verbose=2, validation_data=(X_val, y_val))

Model: "sequential_5"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding_5 (Embedding)     (None, 30, 50)            1350      
                                                                 
 gru_5 (GRU)                 (None, 150)               90900     
                                                                 
 dense_5 (Dense)             (None, 27)                4077      
                                                                 
Total params: 96,327
Trainable params: 96,327
Non-trainable params: 0
_________________________________________________________________
None
Epoch 1/100
199/199 - 15s - loss: 2.7371 - acc: 0.2164 - val_loss: 2.3978 - val_acc: 0.3286 - 15s/epoch - 76ms/step
Epoch 2/100
199/199 - 12s - loss: 2.2818 - acc: 0.3235 - val_loss: 2.2142 - val_acc: 0.3484 - 12s/epoch - 62ms/step
Epoch 3/100
199/199 - 12s - loss: 2.1493 - acc: 0.3612 - val_loss: 2.1446 - v

Here, we are creating modek using the layers.
first we are creating the layers of model for language model at the last we are fitting the model as well as validation data.

We are using the keras to embedding layer to learn 50 dimension embedding for each character. This will help model understand complex relation between character.We are here using the GRU layer as the base model with 150 timesteps. In conclude, here we are using the final layer as a dense layer with a softmax activation for prediction.

We are here providing the multiple layers such as embedding, GRU, Dense layers. after the 100 epochs of the text and we are getting the accuracy and loss as well as the evaluation score of the language model.

Talking about the epochs 100 the final accuracy is 0.90 which is much acceptable while loss is 0.28 which is good ration and its shows the better performace of the language model.after that we are evaluating the model and calculating the score for the model which is validation accuracy is 0.47 which quite acceptable and better while the validation loss is 3.41 which is higher than what we got while generating model. so, all in all the neural language model is better in performace and accurate to predict the next word in the word sequences.

**Step-8** Now model is created now we are input the text into the model and creating new text.

In [ ]:

# generate a sequence of characters with a language model
def generate_seq(model, mapping, seq_length, seed_text, n_chars):
	in_text = seed_text
	# generate a fixed number of characters
	for _ in range(n_chars):
		# encode the characters as integers
		encoded = [mapping[char] for char in in_text]
		# truncate sequences to a fixed length
		encoded = pad_sequences([encoded], maxlen=seq_length, truncating='pre')
		#predict character
		yhat=model.predict(encoded)
		yhat=np.argmax(yhat,axis=1)
		#yhat = model.predict_classes(encoded, verbose=0)
		# reverse map integer to character
		out_char = ''
		for char, index in mapping.items():
			if index == yhat:
				out_char = char
				break
		# append to input
		in_text += char
	return in_text

inp='large armies for'
print(len(inp))
print(generate_seq(model, mapping, 30, inp.lower(),15))

16
1/1 [==============================] - 0s 28ms/step
large armies foreign our consti


Here, we are using the created neural language model and now we are inputing the text into the model and generating new text of sequences.

Here we are inserting the text into the model to check the model working condition. at the last we provide some words and we need to get the sequences of the words after that.

Here we input the text into neural language model and check whether its working perfectly or not and based on the results its working perfectly and predicting the next word of the sequence is quite accurate. we also calculate the length of the input to get the better result or we can say better prediction of next word of the sequences from the model.

In conclude, its proven that neural language model is more accurate and better performer for language model and as well as for achiving the goal for machine translation neural language model is accurate.

#Task 4 (10 points)
What are the limitations of N-gram language modeling in natural language processing? Provide an example of a situation where N-gram language modeling may not be effective.

**Answer here:**
N-gram language model is the most popular technique in neural language processing that predict the probability of the next word based on the condition n-1 word.

Talking about the limitation of the N-gram language modeling in natural language processing.

Sparsity : N-gram model is suffering from the sparsity issues when dealing with rare or unseen n-gram. when the n-grams increases the probabilities of encountering unseen n-gram increase and finally result the inaccurate predictions.

Ambiguity :  n-gram doesn't handle ambiguity effectively.In sequence of words with same probability where the multiple words follows the sequence of words nut model may face the issue to choose the correct sequence one.

limited context: n-hram models only consider a limited number of previous words in the sentences, its typically range between 1 word to 5 words, that means may be that don't able to capture the long dependent words between words and result the inaccurate pridictions.

For instance of if there is a situation where N-gram language model may not able to effective in machine translation. machine translation sometimes meaning of sentence can change based on the order of the words.n-gram model may be not able to work because of the its limited context and ambiguity.

Talking about the example, if the sentence is " I am happy to see you" or another sentence is "I am glad that you can see" So, the correct translation based on the context and the meaning of intention, which may be N-gram model may not able to identify correctly.

In conclude, N-gram model is good with the short sentences where we can define the words and predict the next word based on the previous words. but its also have some limitations which is we discuss in above the main limitation is N-gram does not able to work good with the machine translation.

# Task 5 (10 points)
"What is the difference between an n-gram model and a neural language model in natural language processing, and which one is better for text generation?"

**Answer here:**
N-gram model : this is probability model that trained using the corpus text this is basically used for predicting the probability of the text based on the previous words or counting the words and predict the probability of next word.this i model is used for speech recognition and machine translation as well as predict the text.

Neural language model : neural language model have the capacity to learn the distributed representations, which allow to mitigate the issue of curse of dimensionality because neural network have the ability to effect learn and represent complex pattern and relation between the data.

A neural network language model is a language model based on Neural Networks , exploiting their ability to learn distributed representations to reduce the impact of the curse of dimensionality.

N- gram models and neural language models both are used in natural language processing for text generation.However, they both has a different kind of approach for language models.

N-gram models are a tyoe of statistical language model that will predict the probability of the next word in sequences based on condition n-1 words. For instance, Bigram model can predict the probability of the words based on the previous words in the sentence while n-gram is easy to implement, but its have a limitation context data and may not identify the longe range of the data.

On the other hand, neural language models use artificial neural networks to  do the probability of words in sequences. This model use more layers of neurons to capture information and higher level features for language model.neural network trained in large amount of data and its capture complex relation between sequence of the words and making the better version of the generating natural sounding text.

when talking about the text generation, neural language models are better than n-gram model. because they are generating more diverse and natural text,in contrast, neural languge model also more complex and comparitevily intense, so model may be slower to train and more difficult to implement then n-gram model.

To all in all, while n-gram models are simple and easy to implement, neural language models are more powerful for creating natural sounding text.And on the other hand, the choice between two models its totally depends on the requirements of the nlp tasks.

# Tutorial 2: Perplexity for the N-gram Language Model

Using the test subset of the Reuters corpus, the below code determines the perplexity of a trigram language model.

In n-gram language model, a typical statistic for assessing the effectiveness of language models is perplexity. It gauges how well a linguistic model can foretell a collection of test data. Less perplexity indicates improved test data prediction by the language model.

The likelihood of the words in a test corpus and the probability of the same words predicted by the model are compared in the code below to get the perplexity of an n-gram model. All of the words in the test corpus are lowercased as part of the preprocessing process.

The test corpus's overall word count is kept in the N variable. The probability of the words predicted by the model is computed using the log_prob variable. The loop iterates over the test corpus and computes the probability of the word given the context using the model for each context (a tuple of the two preceding words) and word in the corpus.

The probability's logarithm is added to the log_prob variable if the probability is higher than 0. The logarithm of the probability of a new word is placed in its place if the probability is 0 (i.e., the model has never seen this context and word combination before). Using the equation perplexity = 2**(-log prob / N), the perplexity is finally calculated and reported.

In [ ]:
import math

test_corpus = reuters.sents(categories='interest')
test_corpus = [word.lower() for sent in test_corpus for word in sent]

# Calculate the perplexity
N = len(test_corpus)
log_prob = 0
for i in range(2, N):
    context = (test_corpus[i-2], test_corpus[i-1])
    word = test_corpus[i]
    if context in reutersmodel and reutersmodel[context][word] > 0:
        log_prob += math.log(reutersmodel[context][word], 2)
    else:
        log_prob += math.log(1/N, 2)

perplexity = 2**(-log_prob / N)
print("Perplexity:", perplexity)

Perplexity: 81.44731624070452


Here, basically we are evaluating our language model and calculating the perplexity of the model.

Perplexity : This is used for checking or evaluating the model accuracy that how well that model predicting the samples. this is basically evaluating the model accuracy.

Here,for that we are performing some task as follows,
import math : this is the library which is we are using for performing some mathematic operation for model.

now after that we create a object for reuters text with parameter categories as a interest.after that we are converting text into the lower case.

after all that we start begining of the calculating the perplexity. where first we provide the length of the test corpus. set constant value for log probabilty as 0. and the we are going to use for loop for range start with 2 up to the length of the test corpus text.here in for loop we are checking the probabilty based on the context words and mathematical operation where the condition is log proability 1/ N ,2 where N is the length of the test corpus text and 2 is the starting point of the loop or we can say starting word position of the loops. at the last we are calculating the perplexity where 2**(-log_prob/N) which means that 2 multiply with -log probabilty and devide by the length of the test corpus and at the last we are calculating the perplexity and print it. final perplexity with reuters corpus text is 81.44.The perplexity is lower than 100 its 81.44 it means the model is accurate and its predict the good reuslt or generate accurate text.

In conclusion, we are evaluating the model and for that we are calculating the perplexity of the model using the perplexity calculation using the mathematics use.takling about the evaluation process of the model here we are checking that our model is generating the correct sample text or not using our basic language model.Finally, after all the calculation and after performing the operation we got perplexity is 81.44. It shows that model is very good in prediction with this model or we can say its more accurate because the perplexity count is under 100.

# Task 6 (25 points)

Using the Language model you created in tutorial 1, compute the perplexity for the same model. You can use the same dataset used in task 2. Provide the explanations for your method in detail.

In [ ]:
import math
test_corpus = gutenberg.sents()

test_corpus = [word.lower() for sent in test_corpus for word in sent]

# Calculate the perplexity
N = len(test_corpus)
log_prob = 0
for i in range(2, N):
    context = (test_corpus[i-2], test_corpus[i-1])
    word = test_corpus[i]
    if context in gutenbergmodel and gutenbergmodel[context][word] > 0:
        log_prob += math.log(gutenbergmodel[context][word], 2)
    else:
        log_prob += math.log(1/N, 2)

perplexity = 2**(-log_prob / N)
print("Perplexity:", perplexity)

Perplexity: 295.4029482623658


**Answer here:**
Here, basically we are evaluating our language model and calculating the perplexity of the model.

Perplexity : This is used for checking or evaluating the model accuracy that how well that model predicting the samples. this is basically evaluating the model accuracy.

Here,for that we are performing some task as follows, import math : this is the library which is we are using for performing some mathematic operation for model.

now after that we create a object for gutenberg text and object name as test_corpus.after that we are converting text into the lower case.

after all that we start begining of the calculating the perplexity. where first we provide the length of the test corpus. set constant value for log probabilty as 0. and the we are going to use for loop for range start with 2 up to the length of the test corpus text.here in for loop we are checking the probabilty based on the context words and mathematical operation where the condition is log proability 1/ N ,2 where N is the length of the test corpus text and 2 is the starting point of the loop or we can say starting word position of the loops. at the last we are calculating the perplexity where 2**(-log_prob/N) which means that 2 multiply with -log probabilty and devide by the length of the test corpus and at the last we are calculating the perplexity and print it. final perplexity with gutenberg corpus text is 295.44.The perplexity is higher than 100 its 295.44. it means the model is not accurate and its not able to predict the good result or generate accurate text.

In conclusion, we are evaluating the model and for that we are calculating the perplexity of the model using the perplexity calculation using the mathematics use.takling about the evaluation process of the model here we are checking that our model is generating the correct sample text or not using our basic language model.Finally, after all the calculation and after performing the operation we got perplexity is 295.44. It shows that model is not very good in predicting with this model or we can say its not accurate.

# Task 7 (5 points)
Lower perplexity generally means better-performing language model. Do you agree with the statement? Explain why do you think so.

**Answer here:**

Yes, I agree with the statement that lower perplexity generally means a better-performing language model.

perplexity : this is calculating how the language model predicts the sequences of the words well and its give accurate result of the next prediction of the sequence of the words.talking in the simple way its phenomenon to see the result of language model prediction for next word for sequence of the words.If the perplexity is law then model is not work properly for the next word and its better its prediction.perplexity is majorly access because of the simplicity and effectiveness.

Talking about the language model, the main aim is to get the lower perplexity of a model for given dataset. if the perplexity is law then its define that model is good with sequence of the next words.

Lower perplexity refer to a language modal is capable to predict the future next words accurately.this will be pursued by reducing the uncertainty of the next word in the sequence of the words and that will lower the perplexity score.if the lower perplexity score then model is good in performing task. I agree with the statement that lower perplexity is shows that model is good with performance.

A language model with a lower perplexity is to have better trained and represent of the language and underlying the patterns of the data.this type of model can identify the context and dependencies of two words in the sentence and finally after that results are more accurate for the next word.

For instance, if we have two different language model such as Model A and Model B and we evaluate the models based on the given text. now, we are calculating the perplexity of both models, its indicates that Model A has a lower perplexity compared to Model B. its  implies that Model A is good at predicting the next words in a sentences and its has a good ability to understand the language underlying structure. so, focusing on the perplexity score we can conclude that Model A has a good performace than Model B.

In summarize, lower perplexity is the symbol of better performance of the language model because they are good at predicting the next word in a sequyence of words and trained for better representation of the underlying language patterns.